packages/eslint-plugin-query/src/rules/exhaustive-deps.rule.ts

```diff
@@ -0,0 +1,173 @@
+import type { JSONSchema, TSESLint, TSESTree } from '@typescript-eslint/utils'
+import { AST_NODE_TYPES, ESLintUtils } from '@typescript-eslint/utils'
+import { ExtraUtils } from '../extra-utils'
+
+const messages = {
+  missingDeps:
+    'The following dependencies are missing in your queryKey: {{deps}}',
+}
+
+const ruleSchema: JSONSchema.JSONSchema4 = {
+  type: 'array',
+  minItems: 0,
+  maxItems: 1,
+  items: {
+    type: 'object',
+    properties: {
+      whitelist: {
+        type: 'array',
+        items: {
+          type: 'string',
+        },
+      },
+    },
+  },
+}
+
+const defaultRuleOptions = { whitelist: ['api', 'fetch', 'axios'] }
+
+type RuleMessage = keyof typeof messages
+type RuleOptions = { whitelist?: string[] }
+type RuleContext = Readonly<TSESLint.RuleContext<RuleMessage, [RuleOptions]>>
+
+const createRule = ESLintUtils.RuleCreator(
+  () => 'https://github.com/tanstack/query',
+)
+
+export const exhaustiveDepsRule = createRule<[RuleOptions], RuleMessage>({
+  name: 'exhausitive-deps',
+  meta: {
+    docs: {
+      description:
+        'This rule ensures that all dependencies are passed to the useQuery queryKey parameter',
+      recommended: 'warn',
+    },
+    messages: messages,
+    type: 'problem',
+    fixable: 'code',
+    schema: ruleSchema,
+  },
+  defaultOptions: [defaultRuleOptions],
+  create(context) {
+    return {
+      CallExpression(node) {
+        if (
+          ExtraUtils.isIdentifier(node.callee) &&
+          node.callee.name === 'useQuery'
```

Comment:
instead of checking for `useQuery`, would it be possible to check for `queryKey` and `queryFn`? In the future (v5), when we only allow objects, I think this pattern will be popular:

```
const todoQueries = {
  list: () => ({ queryKey: ['todos'], queryFn: fetchTodos })
  detail: (id) => ({ queryKey: ['todos', id], queryFn: () => fetchTodo(id) })
}

useQuery(todoQueries.list())
useQuery(todoQueries.detail(5))

queryClient.prefetchQuery(todoQueries.detail(23))
```

this kind of takes the "query key factory" towards a full blown "query factory" where you can share the resulting object between `useQuery` and methods like `prefetchQuery` or even `invalidateQueries`...

Reply:
That is a very interesting approach! I'll write some tweaks.